File: react-native/client/android/src/main/java/expo/modules/coinbasewalletsdkexpo/records/WalletRecord.kt
package expo.modules.coinbasewalletsdkexpo

import com.coinbase.android.nativesdk.Wallet
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class WalletIdentifierRecord : Record {
    @Field
    var platform: String = "android"

    @Field
    var packageName: String = ""
}

class WalletRecord : Record {
    @Field
    var name: String = ""

    @Field
    var iconUrl: String = ""

    @Field
    var url: String = ""

    @Field
    var appStoreUrl: String = ""

    @Field
    var id: WalletIdentifierRecord = WalletIdentifierRecord()

}

val Wallet.asRecord: WalletRecord
    get() {
        val id = WalletIdentifierRecord()
        id.platform = "android"
        id.packageName = this.packageName

        val record = WalletRecord()
        record.name = this.name
        record.iconUrl = this.iconUrl
        record.url = this.url
        record.id = id
        return record
    }

val WalletRecord.asWallet: Wallet
    get() {
        return Wallet(
            name = this.name,
            iconUrl = this.iconUrl,
            url = this.url,
            packageName = this.id.packageName
        )
    }

File: react-native/client/ios/Records/WalletRecord.swift
import CoinbaseWalletSDK
import ExpoModulesCore
import Foundation

struct WalletIdentifierRecord : Record {
    @Field
    var platform: String = "ios"

    @Field
    var mwpScheme: String
}

struct WalletRecord : Record {
    @Field
    var name: String

    @Field
    var iconUrl: String

    @Field
    var url: String

    @Field
    var appStoreUrl: String

    @Field
    var id: WalletIdentifierRecord
}

extension WalletRecord {
    var asWallet: Wallet {
        return Wallet(
            name: self.name,
            iconUrl: URL(string:  self.iconUrl)!,
            url: URL(string: self.url)!,
            mwpScheme: URL(string: self.id.mwpScheme)!,
            appStoreUrl: URL(string: self.appStoreUrl)!
        )
    }
}

extension Wallet {
    var asRecord: WalletRecord.Dict {
        let id = WalletIdentifierRecord()
        id.mwpScheme = self.mwpScheme.absoluteString

        let record = WalletRecord()
        record.name = self.name
        record.iconUrl = self.iconUrl.absoluteString
        record.url = self.url.absoluteString
        record.appStoreUrl = self.appStoreUrl.absoluteString
        record.id = id
        return record.toDictionary()
    }
}
